### PR TITLE
Add object storage

### PR DESCRIPTION
**Pull request by junjiejiangjjj**

### CODE REVIEW

File: gptcache/manager/object_data/local_storage.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import uuid
+from pathlib import Path
+from gptcache.manager.object_data.base import ObjectBase
+
+
+class LocalObjectStorage(ObjectBase):
+    """Local object storage
+    """
+
+    def __init__(self, local_root):
```
Comment (SimFG) [inline gptcache/manager/object_data/local_storage.py:12]: add param type, like: `local_root: str`

File: gptcache/manager/object_data/local_storage.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    def get_access_link(self, obj: str, _: int = 3600):
+        return obj
+
+    def delete(self, to_delete: List[str]):
+        assert isinstance(to_delete, list)
+        for obj in to_delete:
+            try:
+                os.remove(obj)
+            except Exception:  # pylint: disable=broad-except
+                pass
```
Comment (SimFG) [inline gptcache/manager/object_data/local_storage.py:37]: give some log

File: gptcache/manager/object_data/manager.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    def __init__(self):
+        raise EnvironmentError(
+            "CacheBase is designed to be instantiated, please using the `CacheBase.get(name)`."
+        )
+
+    @staticmethod
+    def get(name, **kwargs):
+        if name == "local":
+            from gptcache.manager.object_data.local_storage import LocalObjectStorage  # pylint: disable=import-outside-toplevel
+            object_base = LocalObjectStorage(kwargs.get("path", "./"))
```
Comment (SimFG) [inline gptcache/manager/object_data/manager.py:18]: the default value `./local_obj` may be more suitable

File: gptcache/manager/object_data/manager.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        elif name == "s3":
+            from gptcache.manager.object_data.s3_storage import S3Storage  # pylint: disable=import-outside-toplevel
+            object_base = S3Storage(kwargs.get("path_prefix"), kwargs.get("bucket"),
+                                    kwargs.get("access_key"), kwargs.get("secret_key"))
```
Comment (SimFG) [inline gptcache/manager/object_data/manager.py:22]: need `endpoint` param

File: gptcache/manager/scalar_data/base.py [hunk 5]
```diff
@@ -13,16 +13,40 @@ class AnswerType(IntEnum):
...
+
+@dataclass
+class CacheData:
+    """
+    CacheData
+    """
+
     question: Any
-    answer: Union[Any, List[Any]]
+    answer: List[Answer]
```
Comment (SimFG) [inline gptcache/manager/scalar_data/base.py:34]: rename it, `answers`

File: gptcache/adapter/adapter.py [hunk 6]
```diff
@@ -49,22 +49,21 @@ def adapt(llm_handler, cache_data_convert, update_cache_callback, *args, **kwarg
...
             rank = chat_cache.similarity_evaluation.evaluation(
                 {
                     "question": pre_embedding_data,
                     "embedding": embedding_data,
                 },
                 {
-                    "question": cache_question,
-                    "answer": cache_answer,
+                    "question": ret.question,
+                    "answer": ret.answer[0].answer,
```
[the remark this answers is not in the document]
Comment (junjiejiangjjj) [reply gptcache/adapter/adapter.py]: fixed

File: gptcache/adapter/adapter.py [hunk 7]
```diff
@@ -49,22 +49,21 @@ def adapt(llm_handler, cache_data_convert, update_cache_callback, *args, **kwarg
...
                     "search_result": cache_data,
-                    "embedding": cache_embedding
+                    "embedding": ret.embedding_data
                 },
                 extra_param=context.get("evaluation_func", None),
             )
             if rank_threshold <= rank:
-                cache_answers.append((rank, cache_answer))
+                cache_answers.append((rank, ret.answer[0].answer))
```
[the remark this answers is not in the document]
Comment (junjiejiangjjj) [reply gptcache/adapter/adapter.py]: fixed

File: gptcache/manager/data_manager.py [hunk 8]
```diff
@@ -180,10 +188,16 @@ def import_data(
             normalize(embedding_data) for embedding_data in embedding_datas
         ]
         for i, embedding_data in enumerate(embedding_datas):
+            if not isinstance(answers[i], str):
```
[the remark this answers is not in the document]
Comment (junjiejiangjjj) [reply gptcache/manager/data_manager.py]: fixed

File: gptcache/manager/data_manager.py [hunk 9]
```diff
@@ -167,11 +165,21 @@ def save(self, question, answer, embedding_data, **kwargs):
 
         if self.cur_size >= self.max_size:
             self._clear()
-
         self.import_data([question], [answer], [embedding_data])
 
+    def _process_answers(self, answers: Union[Answer, List[Answer]]):
```
[the remark this answers is not in the document]
Comment (junjiejiangjjj) [reply gptcache/manager/data_manager.py]: fixed

File: gptcache/manager/object_data/local_storage.py (hunk 1 above)
Comment (junjiejiangjjj) [reply gptcache/manager/object_data/local_storage.py]: fixed

File: gptcache/manager/object_data/local_storage.py (hunk 2 above)
Comment (junjiejiangjjj) [reply gptcache/manager/object_data/local_storage.py:42]: fixed

File: gptcache/manager/object_data/manager.py (hunk 3 above)
Comment (junjiejiangjjj) [reply gptcache/manager/object_data/manager.py]: fixed

File: gptcache/manager/object_data/manager.py (hunk 4 above)
Comment (junjiejiangjjj) [reply gptcache/manager/object_data/manager.py]: fixed

File: gptcache/manager/scalar_data/base.py (hunk 5 above)
Comment (junjiejiangjjj) [reply gptcache/manager/scalar_data/base.py]: fixed

**Pull request merged**